### ISSUE TITLE
Sync Google Directory contacts

### ISSUE THREAD

**Issue by gschwim** [zanonmark/Google-4-TbSync#33]
Hello,

I've noticed with GSuite/Workspace accounts, contacts in the "Directory" subsection do not appear to be able to sync. Currently a workaround for this is to copy the entire Directory into your contacts manually.

Is there any capability or future plan to directly sync down the Directory? If it is currently able to be done I've not been able to find any docs that support setting this up.

**Comment by gschwim** [comment]
Sure!

**Comment by zanonmark** [comment]
Thanks, I will work on it as soon as I find a little time... I think it will just take some small copy&paste work on your side...

Thanks,
MZ

**Comment by theblackhole** [comment]
Hi,
Any news on this feature ? :)

**Comment by zanonmark** [comment]
Hi,

no sorry, I've been very busy...

I'm reading https://developers.google.com/people/api/rest/v1/people/listDirectoryPeople.
Could You please add the following OAuth scope https://www.googleapis.com/auth/directory.readonly to Your Google Cloud project, so that we can start making some tests using Your account?

Also, how would You like those directory contacts to be imported? as normal contacts? in a specific contact group (aka mailing-list)?

Thanks,
MZ

**Comment by theblackhole** [comment]
Hi @zanonmark and happy new year ! :)

No worries, you dont have to be sorry, we all have stuff to do :)

I added the scope and ready to make some tests.

I think a "Directory" contact group would be the best as it would match Google Contacts UX :

![image](https://user-images.githubusercontent.com/7818904/211774586-d8be5caf-6ec3-48ed-8911-8c54c1d5815b.png)

In the end, it could look like that :

![image](https://user-images.githubusercontent.com/7818904/211776145-caf5f9c7-442e-4cef-ab51-155b2c151dec.png)

I don't know if you can make this contact group special (another icon/color... in my POC I added an underscore so it would stay on top), if you can that would be a great bonus, if not as long as directory contacts are available for autocompletion in the "To:" field :)

Thanks

**Comment by MSBlueVn** [comment]
I am also looking for a solution for this, I work under a Google Workspace account, and want the addresses to autocomplete from the directory.
Is there anything I can help with to progress this issue?

**Comment by zanonmark** [comment]
Hi,

the underlying engine is being heavily rewritten to make use of the new WebExtension API.

Once this next version is published, I'll be working on adding new features like this one.
At that time I will certainly need some help, because I don't have a Google Workspace account to test with. So I will email You a -dev version and You will kindly run it on Your PCs... :)

Thanks,
MZ

**Comment by MSBlueVn** [comment]
I'll be happy to help.

**Comment by MSBlueVn** [comment]
@zanonmark Any way I can help with this? I really hate opening Gmail every time I need to send to someone I haven't sent before.

**Comment by ryanleesipes** [comment]
Why not offer a patch to Thunderbird itself? We'd accept.

**Comment by zanonmark** [comment]
Just an update to this ticket.
In 2025 Google-4-TbSync will be rewritten to be fully WebExtension-based and will work on top of TB itself (not TbSync, which will be merged into TB).
At that time I will consider adding this - and other - enhancements to the engine.

Thanks,
MZ

**Comment by MSBlueVn** [comment]
@zanonmark It has been a year since the last comment any idea if this will happen in 2025?

**Comment by axyklee** [comment]
@zanonmark I know that you are planning to rewrite this as a WebExt directly on top of TB, but here's a relatively easy fix for now. You mention not having a Workspace account, so additional testers might be needed.

**Comment by MSBlueVn** [comment]
I will be willing to test, if this means getting contact suggestions and autocomplete from my Google Workspace

**Comment by axyklee** [comment]
@MSBlueVn You can already test this with my PR. Either compress the files into a `.zip` file and change the extension to `.xpi` or select the `manifest.json` in debugging mode.
I would compile the `.xpi` for you, but it seems like GitHub doesn't like this file type.

Here's a compressed `.zip`. Change the file extension to `.xpi` and it should work!

[Google-4-TbSync.zip](https://github.com/user-attachments/files/23768084/Google-4-TbSync.zip)

**Comment by MSBlueVn** [comment]
Thanks, this works!

**Comment by MSBlueVn** [comment]
I recently got hangs in thunderbird, here's the debug for when it does:
`r: can't access property "replace" of undefined
console.error: WebExtensions: 
  Message: Error: An unexpected error occurred
  Stack:
    normalizeError@resource://gre/modules/ExtensionCommon.sys.mjs:820:12
wrapPromise/</<@resource://gre/modules/ExtensionCommon.sys.mjs:970:23

console.error: "[calCachedCalendar] replay action failed: <unknown>, uri=googleapi://m.s@e.com/?calendar=m.s%40e.com, result=An unexpected error occurred, operation=[object Object]"
console.log: "AddressBookSynchronizer.synchronize(): Read-only mode detected."
console.log: "AddressBookSynchronizer.synchronize(): Preparing the target address book item map."
console.log: "AddressBookSynchronizer.synchronize(): Preparing the contact group member map."
console.log: "AddressBookSynchronizer.synchronize(): Retrieving all the local changes since the last synchronization."
console.log: "AddressBookSynchronizer.synchronize(): Synchronization started."
console.log: "AddressBookSynchronizer.synchronizeContacts(): Cycling on the server contacts."
console.log: "AddressBookSynchronizer.synchronizeContacts(): Cycling on the locally added contacts."
console.log: "AddressBookSynchronizer.synchronizeContacts(): Cycling on the locally modified contacts."
console.log: "AddressBookSynchronizer.synchronizeContacts(): Determining all the remotely deleted contacts."
console.log: "AddressBookSynchronizer.synchronizeContactGroups(): Cycling on the server contact groups."
console.log: "AddressBookSynchronizer.synchronizeContactGroups(): Cycling on the locally added contact groups."
console.log: "AddressBookSynchronizer.synchronizeContactGroups(): Cycling on the locally modified contact groups."
console.log: "AddressBookSynchronizer.synchronizeContactGroups(): Determining all the remotely deleted contact groups."
console.log: "AddressBookSynchronizer.synchronizeContactGroupMembers(): Determining all the members for each contact group."
console.log: "AddressBookSynchronizer.synchronize(): Fixing the change log."
console.log: "AddressBookSynchronizer.synchronize(): Synchronization finished."
console.log: "AddressBookSynchronizer.synchronize(): Read-only mode detected."
console.log: "AddressBookSynchronizer.synchronize(): Preparing the target address book item map."
console.log: "AddressBookSynchronizer.synchronize(): Preparing the contact group member map."
console.log: "AddressBookSynchronizer.synchronize(): Retrieving all the local changes since the last synchronization."
console.log: "AddressBookSynchronizer.synchronize(): Synchronization started."
console.log: "AddressBookSynchronizer.synchronizeContacts(): Cycling on the server contacts."
console.log: "AddressBookSynchronizer.synchronizeContacts(): Cycling on the locally added contacts."
console.log: "AddressBookSynchronizer.synchronizeContacts(): Cycling on the locally modified contacts."
console.log: "AddressBookSynchronizer.synchronizeContacts(): Determining all the remotely deleted contacts."
console.log: "AddressBookSynchronizer.synchronizeContactGroups(): Cycling on the server contact groups."
console.log: "AddressBookSynchronizer.synchronizeContactGroups(): Cycling on the locally added contact groups."
console.log: "AddressBookSynchronizer.synchronizeContactGroups(): Cycling on the locally modified contact groups."
console.log: "AddressBookSynchronizer.synchronizeContactGroups(): Determining all the remotely deleted contact groups."
console.log: "AddressBookSynchronizer.synchronizeContactGroupMembers(): Determining all the members for each contact group."
console.log: "AddressBookSynchronizer.synchronize(): Fixing the change log."
console.log: "AddressBookSynchronizer.synchronize(): Synchronization finished."
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
JavaScript error: resource:///modules/calendar/Ical.sys.mjs, line 7221: TypeError: can't access property "replace", val is undefined
console.error: WebExtensions: 
  Message: Error: An unexpected error occurred
  Stack:
    normalizeError@resource://gre/modules/ExtensionCommon.sys.mjs:820:12
wrapPromise/</<@resource://gre/modules/ExtensionCommon.sys.mjs:970:23

console.error: "[calCachedCalendar] replay action failed: <unknown>, uri=googleapi://m.s@e.com/?calendar=m.s%40e.com, result=An unexpected error occurred, operation=[object Object]"

`

any way we can fix this?

**Comment by axyklee** [comment]
@MSBlueVn Hi, sorry to get back to you so late. Can I confirm that the hangs did not occur when you first used the extension and only randomly occurs after you used it for a while? How many (approx) Google Directory contacts do you have?

I must admit, I have noticed that my uni's ~1000 people Directory does cause major hangs when first syncing, but it seems fine afterwards.

However, it seems like the logs have nothing to do with my PR, since I did not make use of the `replace()` function nor have utilized any variable named `val`. The error also suggests something to do with parsing calendars, which is outside the scope of this PR.

**Comment by MSBlueVn** [comment]
This was causing my thunderbird to lag and crash, so I gave up and moved to 
CardBook
the integration there works out of the box.

**Comment by axyklee** [comment]
Good to hear that you found a solution. Do directory contacts also work on CardBook?

**Comment by MSBlueVn** [comment]
Yes, that's what I use it for, finding directory contacts when I type a name in the to fields.

**Comment by axyklee** [comment]
@MSBlueVn I'm sorry, is [this extension](https://services.addons.thunderbird.net/zh-Tw/thunderbird/addon/cardbook/) the one you are talking about? I don't see my directory contacts being synced. Is there a specific setting that I need to enable?

**Comment by MSBlueVn** [comment]
Yes that's the one. here's what you need to do:
Step-by-Step Configuration
Open CardBook: Click the CardBook icon in the top right of the toolbar or go to Tools > CardBook.
Add New Address Book:
Right-click on "Address Books" in the left-hand panel, or click the 3-bar icon (menu) at the top left and select Address Book -> New Address Book.
Select Remote: Select Remote and click Next.
Select CardDAV: Select CardDAV.
Configure Account:
User Name: Enter your full Google Workspace email address.
Location: Keep this blank. CardBook often automatically detects the correct directory URL when using Google OAuth.
Click Validate.
Authentication:
A Google OAuth2 window will pop up. Enter your password and allow Thunderbird/CardBook to access your contacts.
If validation fails, try entering https://googleapis.com in the location field.
Finalize: Once validated, click Next, then click OK to complete the setup. 

Accessing the Directory
Once synced, your directory contacts will appear in the new address book created in the CardBook tab. 
Important: Note that Google Directory contacts are "read-only" within third-party apps like Thunderbird.
Autocomplete: To use these contacts, ensure CardBook is configured to share its address books with Thunderbird for searching in the compose window.

**Comment by axyklee** [comment]
Ah yes thank you for your help but your instructions are actually slightly incorrect. Users shouldn't select "CardDAV" but should select "Google Workspace" (note: not "Google")!

**Comment by MSBlueVn** [comment]
Yeah I did this a while back and this was what Google wrote. glad you figured it out.
